packages/core/database/src/migrations/internal-migrations/document-id.ts

```diff
@@ -0,0 +1,170 @@
+      `
+    SELECT :tableName:.id as id, string_agg(DISTINCT :inverseJoinColumn:::character varying, ',') as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :tableName:.id, :joinColumn:
+    LIMIT 1;
+  `,
+      params
+    );
+
+    return res.rows;
+  },
+  async mysql(knex: Knex, params: Params) {
+    const [res] = await knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :tableName:.id, :joinColumn:
+    LIMIT 1;
+  `,
+      params
+    );
+
+    return res;
+  },
+  async sqlite(knex: Knex, params: Params) {
+    return knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :joinColumn:
+    LIMIT 1;
+    `,
+      params
+    );
+  },
+};
+
+const getNextIdsToCreateDocumentId = async (
+  db: Database,
+  knex: Knex,
+  {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  }: {
+    joinColumn: string;
+    inverseJoinColumn: string;
+    tableName: string;
+    joinTableName: string;
+  }
+): Promise<number[]> => {
+  const res = await QUERIES[db.dialect.client as keyof typeof QUERIES](knex, {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  });
+
+  if (res.length > 0) {
+    const row = res[0];
+    const otherIds = row.other_ids
+      ? row.other_ids.split(',').map((v: string) => parseInt(v, 10))
+      : [];
+
+    return [row.id, ...otherIds];
+  }
+
+  return [];
+};
+
+// Migrate document ids for tables that have localizations
+const migrateDocumentIdsWithLocalizations = async (db: Database, knex: Knex, meta: Meta) => {
+  const joinColumn = `${meta.singularName.toLowerCase()}_id`;
+  const inverseJoinColumn = `inv_${meta.singularName.toLowerCase()}_id`;
+
+  let ids: number[];
+
+  do {
+    ids = await getNextIdsToCreateDocumentId(db, knex, {
+      joinColumn,
+      inverseJoinColumn,
+      tableName: meta.tableName,
+      joinTableName: `${meta.tableName}_localizations_links`,
+    });
+
+    if (ids.length > 0) {
+      await knex(meta.tableName).update({ document_id: createId() }).whereIn('id', ids);
+    }
+  } while (ids.length > 0);
+};
+
+// Migrate document ids for tables that don't have localizations
+const migrationDocumentIds = async (db: Database, knex: Knex, meta: Meta) => {
+  let res;
+  do {
+    res = await knex.select('id').from(meta.tableName).whereNull('document_id').limit(1).first();
+
+    if (res) {
+      await knex(meta.tableName).update({ document_id: createId() }).whereIn('id', res.id);
+    }
+  } while (res !== null);
```

Comment:
Maybe the following is too complicated compared to your approach, but an idea to improve the performance of this in the future:

Instead of doing 2 queries per id, we could create a temporary table to store the generated id-docId pairs

So we could batch load the ids first:
```ts
  const rows = await knex.select('id')
                         .from(meta.tableName)
                         .whereNull('document_id')
                         .limit(batchSize);
```

Then store these values with its own ids in a temporary table
```ts
  await knex.schema.createTable('temp_id_updates', table => {
    table.integer('temp_id');
    table.string('temp_document_id');
  });

  await knex('temp_id_updates').insert(idDocumentPairs.map(pair => ({
    temp_id: pair.id,
    temp_document_id: createId()
  })));
```

Then updating the collection type table with these values, and droping the temp table.

Reply:
Just did a small calc of how long would it take to execute this migration assuming a 10ms latency per request:

10k records -> ~3min
50k records -> ~16min
100k records -> ~33min

I think after the public release we could spend a bit of time enhancing the performance of this or providing some logs to point the users the status of the migration

Reply:
Yeah, I think we could start with the current version since it works, but we should review performance improvements immediately after.

Reply:
Yeah I thought about it but you need a looping mechanism for the update query to work and I couldn't make that work cleanly without spending more time on it :)